File: store/src/test/java/com/nytimes/android/external/store/DontCacheErrorsTest.java
package com.nytimes.android.external.store;

import com.nytimes.android.external.store.base.impl.BarCode;
import com.nytimes.android.external.store.base.impl.Store;
import com.nytimes.android.external.store.base.impl.StoreBuilder;

import org.junit.Before;
import org.junit.Test;

import rx.Observable;

public class DontCacheErrorsTest {

    boolean shouldThrow;
    private Store<Integer, BarCode> store;

    @Before
    public void setUp() {
        store = StoreBuilder.<Integer>barcode()
                .fetcher(barCode -> Observable.fromCallable(() -> {
                    if (shouldThrow) {
                        throw new RuntimeException();
                    } else {
                        return 0;
                    }
                }))
                .open();
    }

    @Test
    public void testStoreDoesntCacheErrors() throws InterruptedException {
        BarCode barcode = new BarCode("bar", "code");

        shouldThrow = true;
        store.get(barcode).test()
                .awaitTerminalEvent()
                .assertError(Exception.class);

        shouldThrow = false;
        store.get(barcode).test()
                .awaitTerminalEvent()
                .assertNoErrors();
    }
}

File: app/src/test/java/com/nytimes/android/sample/DontCacheErrorsTest.kt
package com.nytimes.android.sample

import com.nytimes.android.external.store3.base.Fetcher
import com.nytimes.android.external.store3.base.impl.BarCode
import com.nytimes.android.external.store3.base.impl.Store
import com.nytimes.android.external.store3.base.impl.StoreBuilder
import kotlinx.coroutines.runBlocking
import org.junit.Before
import org.junit.Test


class DontCacheErrorsTest {

    internal var shouldThrow: Boolean = false
    lateinit var store: Store<Int, BarCode>

    @Before
    fun setUp() {
        store = StoreBuilder.barcode<Int>()
                .fetcher(object : Fetcher<Int, BarCode> {
                    override suspend fun fetch(key: BarCode): Int {
                        if (shouldThrow) {
                            throw RuntimeException()
                        } else {
                            return 0
                        }
                    }
                })
                .open()
    }

    @Test(expected = java.lang.RuntimeException::class)
    fun testStoreDoesntCacheErrors() {
        runBlocking {
            val barcode = BarCode("bar", "code")
            shouldThrow = true
            val job = store.get(barcode)
        }
    }
}
